src/pip/_internal/resolution/resolvelib/factory.py

```diff
@@ -155,7 +159,12 @@ def make_requirement_from_install_req(self, ireq):
                 ireq.link, extras=set(), parent=ireq,
             )
             return ExplicitRequirement(cand)
-        return SpecifierRequirement(ireq, factory=self)
+
+        specifier = SpecifierRequirement(ireq, factory=self)
+        if ireq.constraint:
+            return ConstraintRequirement(specifier, factory=self)
+
+        return specifier
```

Comment:
I got confused for a while by this name (and the attribute `ConstraintRequirement._specifier`). The name yells `packaging.specifiers.SpecifierSet` but they are totally not 😖

Reply:
Basically it felt like `specifier_requirement` is too long. But I can't think of a reasonable abbreviation (I'd prefer not to use `req` as we use that for an `InstallRequirement` all over the place).

I've changed it to `specifier_requirement`.